File: src/test/java/com/hileco/cortex/analysis/attack/AttackerTest.java
package com.hileco.cortex.analysis.attack;

import com.hileco.cortex.analysis.GraphBuilder;
import com.hileco.cortex.analysis.processors.FlowProcessor;
import com.hileco.cortex.analysis.processors.JumpUnreachableProcessor;
import com.hileco.cortex.analysis.processors.ParameterProcessor;
import com.hileco.cortex.constraints.expressions.Expression;
import com.hileco.cortex.documentation.Documentation;
import com.hileco.cortex.instructions.InstructionsBuilder;
import com.hileco.cortex.instructions.conditions.EQUALS;
import com.hileco.cortex.instructions.debug.HALT;
import com.hileco.cortex.instructions.io.LOAD;
import com.hileco.cortex.instructions.math.DIVIDE;
import com.hileco.cortex.instructions.stack.PUSH;
import org.junit.Assert;
import org.junit.Test;

import java.math.BigInteger;
import java.util.List;

import static com.hileco.cortex.instructions.ProgramException.Reason.WINNER;
import static com.hileco.cortex.vm.ProgramStoreZone.CALL_DATA;

public class AttackerTest {

    @Test
    public void test() {
        var graphBuilder = new GraphBuilder(List.of(
                new ParameterProcessor(),
                new FlowProcessor(),
                new JumpUnreachableProcessor()
        ));
        var instructionsBuilder = new InstructionsBuilder();
        instructionsBuilder.IF(conditionBuilder -> conditionBuilder.include(List.of(new PUSH(BigInteger.valueOf(2).toByteArray()),
                                                                                    new PUSH(BigInteger.valueOf(1).toByteArray()),
                                                                                    new LOAD(CALL_DATA),
                                                                                    new DIVIDE(),
                                                                                    new PUSH(BigInteger.valueOf(12345).toByteArray()),
                                                                                    new EQUALS())),
                               contentBuilder -> contentBuilder.include(List.of(new HALT(WINNER))));
        var instructions = instructionsBuilder.build();
        var graph = graphBuilder.build(instructions);
        var attacker = new Attacker(Attacker.TARGET_IS_HALT_WINNER);
        var solutions = attacker.solve(graph);
        Documentation.of(Attacker.class.getSimpleName())
                .headingParagraph(Attacker.class.getSimpleName())
                .paragraph("Program:").source(instructions)
                .paragraph("Attack method:").source("TARGET_IS_HALT_WINNER")
                .paragraph("Suggested solution by Cortex:").source(solutions);
        Assert.assertEquals(1, solutions.size());
        var solution = solutions.get(0);
        Assert.assertTrue(solution.isSolvable());
        Assert.assertEquals(1, solution.getPossibleValues().size());
        var entry = solution.getPossibleValues().entrySet().iterator().next();
        Assert.assertEquals(new Expression.Reference(CALL_DATA, new Expression.Value(1L)), entry.getKey());
        Assert.assertEquals(Long.valueOf(24690L), entry.getValue());
    }
}

File: src/test/java/com/hileco/cortex/analysis/attack/AttackerTest.kt
package com.hileco.cortex.analysis.attack

import com.hileco.cortex.analysis.BarrierProgram.Companion.BARRIER_00
import com.hileco.cortex.analysis.BarrierProgram.Companion.BARRIER_01
import com.hileco.cortex.analysis.GraphBuilder
import com.hileco.cortex.constraints.expressions.Expression
import com.hileco.cortex.documentation.Documentation
import com.hileco.cortex.vm.ProgramStoreZone.CALL_DATA
import org.junit.Assert
import org.junit.Test

class AttackerTest {
    @Test
    fun testBarrier00() {
        val graph = GraphBuilder.BASIC_GRAPH_BUILDER.build(BARRIER_00.instructions)
        val attacker = Attacker(Attacker.TARGET_IS_HALT_WINNER)
        val solutions = attacker.solve(graph)
        Assert.assertEquals(1, solutions.size)
        val solution = solutions.first()
        Assert.assertTrue(solution.isSolvable)
        Assert.assertEquals(0, solution.possibleValues.size)
    }

    @Test
    fun testBarrier01() {
        val graph = GraphBuilder.BASIC_GRAPH_BUILDER.build(BARRIER_01.instructions)
        val attacker = Attacker(Attacker.TARGET_IS_HALT_WINNER)
        val solutions = attacker.solve(graph)
        Documentation.of(Attacker::class.simpleName!!)
                .headingParagraph(Attacker::class.simpleName!!)
                .paragraph("Program:").source(BARRIER_01.instructions)
                .paragraph("Attack method:").source("TARGET_IS_HALT_WINNER")
                .paragraph("Suggested solution by Cortex:").source(solutions)
        Assert.assertEquals(1, solutions.size)
        val solution = solutions.first()
        Assert.assertTrue(solution.isSolvable)
        Assert.assertEquals(1, solution.possibleValues.size)
        val entry = solution.possibleValues.entries.first()
        Assert.assertEquals(Expression.Reference(CALL_DATA, Expression.Value(1L)), entry.key)
        Assert.assertEquals(24690L, entry.value)
    }
}